mypy tests failing

The [mypy test is producing five errors](https://github.com/sympy/sympy/actions/runs/3414593061/jobs/5682788659). They occurred in #24013 which does not modify these Python files, so I believe something else has changed.

```
sympy/core/evalf.py:1749: error: Incompatible default for argument "eps" (default has type "None", argument has type "Expr")  [assignment]
sympy/core/evalf.py:1750: error: Incompatible default for argument "options" (default has type "None", argument has type "Dict[str, Any]")  [assignment]
sympy/combinatorics/free_groups.py:96: error: Function "Type[Expr]" could always be true in boolean context  [truthy-function]
sympy/printing/smtlib.py:74: error: Incompatible default for argument "settings" (default has type "None", argument has type "Dict[Any, Any]")  [assignment]
sympy/printing/smtlib.py:434: error: Incompatible default for argument "symbol_table" (default has type "None", argument has type "Dict[Any, Any]")  [assignment]
```

These things are not really that hard to fix. Example
```
sympy/core/evalf.py:1749: error: Incompatible default for argument "eps" (default has type "None", argument has type "Expr")  [assignment]
```
That's this line.
https://github.com/sympy/sympy/blob/eeb263cd04fbdd54a2fb245304afb88463f8ddb3/sympy/core/evalf.py#L1749-L1750
The type of `eps` has been declared as `Expr` but the default value is `None` which is not `Expr` so although mypy previously allowed it that's not a valid type hint.

The fix here is probably just declaring the types to be e.g. `Optional[Expr]` rather than `Expr`. It's not hard to test this locally:
```console
$ pip install -U mypy
Requirement already satisfied: mypy in ./38venv/lib/python3.8/site-packages (0.950)
Collecting mypy
  Downloading mypy-0.990-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.9/17.9 MB 11.4 MB/s eta 0:00:00
Requirement already satisfied: typing-extensions>=3.10 in ./38venv/lib/python3.8/site-packages (from mypy) (4.0.1)
Requirement already satisfied: mypy-extensions>=0.4.3 in ./38venv/lib/python3.8/site-packages (from mypy) (0.4.3)
Requirement already satisfied: tomli>=1.1.0 in ./38venv/lib/python3.8/site-packages (from mypy) (1.2.3)
Installing collected packages: mypy
  Attempting uninstall: mypy
    Found existing installation: mypy 0.950
    Uninstalling mypy-0.950:
      Successfully uninstalled mypy-0.950
Successfully installed mypy-0.990

[notice] A new release of pip available: 22.2 -> 22.3
[notice] To update, run: pip install --upgrade pip
$ mypy sympy
sympy/utilities/lambdify.py:779: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
sympy/utilities/lambdify.py:852: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
sympy/utilities/lambdify.py:884: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
sympy/core/evalf.py:1749: error: Incompatible default for argument "eps" (default has type "None", argument has type "Expr")  [assignment]
sympy/core/evalf.py:1749: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
sympy/core/evalf.py:1749: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
sympy/core/evalf.py:1750: error: Incompatible default for argument "options" (default has type "None", argument has type "Dict[str, Any]")  [assignment]
sympy/core/evalf.py:1750: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
...
```
The message even suggests that there is a tool that can fix this automatically:
https://github.com/hauntsaninja/no_implicit_optional

--- a/sympy/combinatorics/free_groups.py
+++ b/sympy/combinatorics/free_groups.py
@@ -93,7 +93,7 @@ def _parse_symbols(symbols):
         return tuple()
     if isinstance(symbols, str):
         return _symbols(symbols, seq=True)
-    elif isinstance(symbols, Expr or FreeGroupElement):
+    elif isinstance(symbols, (Expr, FreeGroupElement)):
         return (symbols,)
     elif is_sequence(symbols):
         if all(isinstance(s, str) for s in symbols):

--- a/sympy/core/evalf.py
+++ b/sympy/core/evalf.py
@@ -1746,8 +1746,9 @@ def N(x, n=15, **options):
     return sympify(x, rational=True).evalf(n, **options)
 
 
-def _evalf_with_bounded_error(x: 'Expr', eps: 'Expr' = None, m: int = 0,
-                              options: OPT_DICT = None) -> TMP_RES:
+def _evalf_with_bounded_error(x: 'Expr', eps: 'Optional[Expr]' = None,
+                              m: int = 0,
+                              options: Optional[OPT_DICT] = None) -> TMP_RES:
     """
     Evaluate *x* to within a bounded absolute error.
 

--- a/sympy/printing/smtlib.py
+++ b/sympy/printing/smtlib.py
@@ -71,7 +71,8 @@ class SMTLibPrinter(Printer):
 
     symbol_table: dict
 
-    def __init__(self, settings: dict = None, symbol_table=None):
+    def __init__(self, settings: typing.Optional[dict] = None,
+                 symbol_table=None):
         settings = settings or {}
         self.symbol_table = symbol_table or {}
         Printer.__init__(self, settings)
@@ -431,7 +432,7 @@ def _auto_assert_smtlib(e: Expr, p: SMTLibPrinter, log_warn=print):
 
 def _auto_infer_smtlib_types(
     *exprs: Basic,
-    symbol_table: dict = None
+    symbol_table: typing.Optional[dict] = None
 ) -> dict:
     # [TYPE INFERENCE RULES]
     # X is alone in an expr => X is bool

